File: p2.java
// Time Complexity : Time complexity would be O(n)
// Space Complexity : Space complexity would be constant O(1)
// Did this code successfully run on Leetcode :Yes

// Any problem you faced while coding this : Nno

// Your code here along with comments explaining your approach
//Using two hashmap approach. One hashmap is used to store key as String s and value as String t while other hashmap is viceversa. If the key is present then check if the value matches. if the value does not match return false

class Solution {
    public boolean isIsomorphic(String s, String t) {
        Map<Character, Character> smap = new HashMap<>();
        Map<Character, Character> tmap = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            char schar = s.charAt(i);
            char tchar = t.charAt(i);
            if (!smap.containsKey(schar)) {
                smap.put(schar, tchar);
            } else if (smap.get(schar) != tchar)
                return false;

            if (!tmap.containsKey(tchar)) {
                tmap.put(tchar, schar);
            } else if (tmap.get(tchar) != schar)
                return false;

        }

        return true;

    }
}

File: p2.py
#TC= O(n)
#SC = O(1)
# if runs successfully on leetcode: YES

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        dictStoT = {}
        dictTtoS = {}
    

        for i in range(0, len(s)):
            if s[i] not in dictStoT:
                dictStoT[s[i]] = t[i]
            else:
                if dictStoT[s[i]] != t[i]:
                    return False
            
            if t[i] not in dictTtoS:
                dictTtoS[t[i]] = s[i]
            else:
                if dictTtoS[t[i]] != s[i]:
                    return False
        return True